rse-r-package.Rmd

```diff
@@ -0,0 +1,590 @@
+so we can try running "Build...Install and Restart"
+to build our package,
+re-start our R session (so that memory is clean),
+and load our newly-created package.
+It works!
+
+But there's still that warning.
+To fix it,
+we add this to our roxygen2 comment block:
+
+```{r eval=FALSE}
+#' @importFrom rlang .data
+```
+
+and then modify the calls that use naked column names to be:
+
+```{r eval=FALSE}
+tally_words <- function(all_words) {
+  dplyr::tibble(word = all_words) %>%
+    dplyr::group_by(.data$word) %>%
+    dplyr::tally()
+}
+```
+
+`.data` is a pronoun that allows us to be explicit when we refer to an object inside the data.
+Adding this---i.e.,
+being explicity that `word` is a column of `.data` rather than an undefined variable---finally
+gives us a clean build.
+
+### Exercise: Document dependencies
+
+1.  Modify the functions in the CO2 package to use `package::function` name for everything.
+2.  Modify the `DESCRIPTION` file to document the package's dependencies.
+
+## How can I add data to a package? {#rse-r-package-data}
+
+But we still need to add our cleaned-up data to our package
+and document the package as a whole.
+Working at the console,
+let's create a very small data set:
+
+```{r eval=FALSE}
+words <- c("some", "words", "appear", "some", "times")
+small_data <- tally_words(words)
+small_data
+```
+```
+# A tibble: 4 x 2
+  word       n
+  <chr>  <int>
+1 appear     1
+2 some       2
+3 times      1
+4 words      1
+```
+
+Next,
+we load the `usethis` package with `library(usethis)`
+and called `usethis::use_data(small_data)`
+to store the tibble in `data/small_data.rda`.
+(We could just call `save` with the appropriate parameters,
+but `usethis` provides a cleaner interface.)
+Note that we *must* save the data as `.rda`, not as (for example) `.rds` or `.csv`;
+only `.rda` will be automatically loaded as part of the project.
+
+When we run `Check`,
+we get a complaint about an undocumented data set,
+so we create a file called `R/small_data.R` to hold documentation about the dataset
+and put this in it:
+
+```{r eval=FALSE}
+#' Sample word frequency data.
+#'
+#' This small dataset contains word frequencies for tutorial purposes.
+#'
+#' @docType data
+#'
+#' @format A data frame
+#' \describe{
+#'   \item{word}{The word being counted (chr)}
+#'   \item{n}{The number of occurrences (int)}
+#' }
+"small_data"
+```
+
+Everything except the last line is a roxygen2 comment block
+that describes the data in plain language,
+then uses some tags and directives to document its format and fields.
+The line `@docType` tells roxygen2 that this comment describes data rather than a function,
+and the last line is the string `"small_data"`,
+i.e.,
+the name of the dataset.
+We will create one placeholder R file like this for each of our datasets,
+and each will have that dataset's name as the thing being documented.
+
+We use a similar trick to document the package as a whole:
+we create a file `R/zipffred.R`
```

Comment:
Better to use:

```
usethis::use_package_doc()
```

to create this file.

Reply:
Added as an exercise.